presto-main/src/main/java/com/facebook/presto/event/query/QueryMonitor.java

```diff
@@ -192,8 +205,22 @@ public void queryCompletedEvent(QueryInfo queryInfo)
             }
 
             Optional<String> plan = Optional.empty();
-            if (queryInfo.getPlan().isPresent()) {
-                plan = Optional.of(objectMapper.writeValueAsString(queryInfo.getPlan().get()));
+            try {
+                if (queryInfo.getOutputStage().isPresent()) {
+                    Session session = queryInfo.getSession().toSession(sessionPropertyManager);
+                    plan = Optional.of(textDistributedPlan(
+                            queryInfo.getOutputStage().get(),
+                            functionRegistry,
+                            new ComposableStatsCalculator(ImmutableList.of()),
+                            new CostCalculatorUsingExchanges(() -> 1),
+                            session,
+                            false));
+                }
+                if (!plan.isPresent() && queryInfo.getPlan().isPresent()) {
+                    plan = Optional.of(objectMapper.writeValueAsString(queryInfo.getPlan().get()));
+                }
+            }
+            catch (Exception ignored) {
```

Comment:
uh. This means this functionality may break silently. E.g. when `ComposableStatsCalculator` rejects empty stat rules list in the constructor.. Please be more specific here. What are actually possible problems you're guarding against?

Reply:
understood. Then at least log the exception, as it's totally unexpected